From 1452de5f69e868d40626a96bfdb051db8db98f8b Mon Sep 17 00:00:00 2001
From: Shaun Tancheff <shaun@tancheff.com>
Date: Sun, 21 Aug 2016 15:51:52 -0500
Subject: [PATCH 1/7] Add support for ZDM detection

This patch adds support for identifying ZDM device mappter targets.

Signed-off-by: Shaun Tancheff <shaun.tancheff@seagate.com>
---
 libblkid/src/Makemodule.am             |  1 +
 libblkid/src/superblocks/superblocks.c |  1 +
 libblkid/src/superblocks/superblocks.h |  1 +
 libblkid/src/superblocks/zdm.c         | 93 ++++++++++++++++++++++++++++++++++
 4 files changed, 96 insertions(+)
 create mode 100644 libblkid/src/superblocks/zdm.c

diff --git a/libblkid/src/Makemodule.am b/libblkid/src/Makemodule.am
index 52d0b0a..ef3d18a 100644
--- a/libblkid/src/Makemodule.am
+++ b/libblkid/src/Makemodule.am
@@ -93,6 +93,7 @@ libblkid_la_SOURCES = \
 	libblkid/src/superblocks/vmfs.c \
 	libblkid/src/superblocks/vxfs.c \
 	libblkid/src/superblocks/xfs.c \
+	libblkid/src/superblocks/zdm.c \
 	libblkid/src/superblocks/zfs.c \
 	\
 	libblkid/src/topology/topology.c \
diff --git a/libblkid/src/superblocks/superblocks.c b/libblkid/src/superblocks/superblocks.c
index 0d970f1..abb3aaa 100644
--- a/libblkid/src/superblocks/superblocks.c
+++ b/libblkid/src/superblocks/superblocks.c
@@ -102,6 +102,7 @@ static const struct blkid_idinfo *idinfos[] =
 	&jmraid_idinfo,
 
 	&bcache_idinfo,
+	&zdm_idinfo,
 	&drbd_idinfo,
 	&drbdmanage_idinfo,
 	&drbdproxy_datalog_idinfo,
diff --git a/libblkid/src/superblocks/superblocks.h b/libblkid/src/superblocks/superblocks.h
index 79dba1a..8df25ee 100644
--- a/libblkid/src/superblocks/superblocks.h
+++ b/libblkid/src/superblocks/superblocks.h
@@ -75,6 +75,7 @@ extern const struct blkid_idinfo nilfs2_idinfo;
 extern const struct blkid_idinfo exfat_idinfo;
 extern const struct blkid_idinfo f2fs_idinfo;
 extern const struct blkid_idinfo bcache_idinfo;
+extern const struct blkid_idinfo zdm_idinfo;
 
 /*
  * superblock functions
diff --git a/libblkid/src/superblocks/zdm.c b/libblkid/src/superblocks/zdm.c
new file mode 100644
index 0000000..aaba06c
--- /dev/null
+++ b/libblkid/src/superblocks/zdm.c
@@ -0,0 +1,93 @@
+/*
+ * Copyright (C) 2015 Shaun Tancheff <shaun@tancheff.com>
+ *
+ * This file may be redistributed under the terms of the
+ * GNU Lesser General Public License.
+ *
+ * Based on code fragments from zdm-tools
+ */
+
+#include <stddef.h>
+#include <stdio.h>
+#include <uuid/uuid.h>
+
+#include "superblocks.h"
+#include "crc32.h"
+
+static const char zdm_magic[] = {
+	0x7a, 0x6f, 0x6e, 0x65, 0x63, 0x44, 0x45, 0x56,
+	0x82, 0x65, 0xf5, 0x7f, 0x48, 0xba, 0x6d, 0x81
+};
+
+struct zdm_super_block {
+	uint32_t crc32;
+	uint32_t reserved;
+	uint8_t magic[ARRAY_SIZE(zdm_magic)];
+	uuid_t  uuid;
+	uint32_t version;     /* 0xMMMMmmpt */
+	uint64_t sect_start;
+	uint64_t sect_size;
+	uint32_t mz_metadata_zones;     /* 3 (default) */
+	uint32_t mz_over_provision;     /* 5 (minimum) */
+	uint64_t zdm_blocks;  /* 0 -> <zdm_blocks> for dmsetup table entry */
+	uint32_t discard;     /* if discard support is enabled */
+	uint32_t disk_type;   /* HA | HM */
+	uint32_t zac_zbc;     /* if ZAC / ZBC is supported on backing device */
+	char label[64];
+	uint64_t data_start;  /* zone # of first *DATA* zone */
+	uint64_t zone_size;   /* zone size in 512 byte blocks */
+	uuid_t uuid_meta;     /* UUID of metadata block device */
+	uint32_t is_meta;     /* true if this is the metadata device */
+	uint32_t io_queue;    /* true if I/O queue should be enabled. */
+};
+typedef struct zdm_super_block zdm_super_block_t;
+
+static uint32_t zdm_crc32(zdm_super_block_t *sblk)
+{
+	uint32_t icrc = sblk->crc32;
+	uint8_t *data = (uint8_t *) sblk;
+	size_t sz = sizeof(*sblk);
+	uint32_t calc;
+
+	sblk->crc32 = 0u;
+	calc = crc32(~0u, data, sz) ^ ~0u;
+	sblk->crc32 = icrc;
+
+	return calc;
+}
+
+
+static int probe_zdm(blkid_probe pr, const struct blkid_idmag *mag)
+{
+	zdm_super_block_t *sblk;
+
+	sblk = blkid_probe_get_sb(pr, mag, struct zdm_super_block);
+	if (!sblk)
+		return errno ? -errno : BLKID_PROBE_NONE;
+
+	if (zdm_crc32(sblk) != le32_to_cpu(sblk->crc32))
+		return BLKID_PROBE_NONE;
+
+	if (blkid_probe_set_uuid(pr, sblk->uuid) < 0)
+		return BLKID_PROBE_NONE;
+
+	return BLKID_PROBE_OK;
+}
+
+const struct blkid_idinfo zdm_idinfo =
+{
+	.name		= "zdm",
+	.usage		= BLKID_USAGE_OTHER,
+	.probefunc	= probe_zdm,
+	.minsz		= 1 << 12,
+	.magics		=
+	{
+		{
+		  .magic = zdm_magic,
+		  .len   = sizeof(zdm_magic),
+		  .kboff = 0,
+		  .sboff = offsetof(struct zdm_super_block, magic)
+		} ,
+		{ NULL }
+	}
+};
-- 
2.9.3

